File: src/emitter/mermaid/node-head.ts
import { VARIABLE_DECLARATION_KIND } from "../../serializer/variable-declaration-kind.js";
import { NODE_KIND } from "../../visual-graph/node-kind.js";
import type { VisualNode } from "../../visual-graph/visual-node.js";
import { escape } from "./escape.js";

export function nodeHead(n: VisualNode): string {
  const name = escape(n.name);
  if (n.isJsxElement) {
    // Mermaid `["..."]` labels require HTML-escaped angle brackets so the
    // parser does not mistake them for syntax; the renderer surfaces them
    // as literal `<` / `>` in the output.
    return `&lt;${name}&gt;`;
  }
  switch (n.kind) {
    case NODE_KIND.FunctionDeclaration:
      return `${name}()`;
    case NODE_KIND.ClassDeclaration:
      return `class ${name}`;
    case NODE_KIND.NamedImportBinding:
      return n.importedName !== n.name ? name : `import ${name}`;
    case NODE_KIND.DefaultImportBinding:
    case NODE_KIND.NamespaceImportBinding:
      return `import ${name}`;
    case NODE_KIND.CatchParameter:
      return `catch ${name}`;
    case NODE_KIND.SyntheticImplicitGlobal:
      return `global ${name}`;
    case NODE_KIND.WriteReference:
      return n.declarationKind === VARIABLE_DECLARATION_KIND.Let
        ? `let ${name}`
        : name;
    case NODE_KIND.SyntheticModuleSource:
      return `module ${name}`;
    case NODE_KIND.SyntheticImportIntermediate:
      return `import ${name}`;
    case NODE_KIND.VarBinding:
      return n.initIsFunction ? `${name}()` : `var ${name}`;
    case NODE_KIND.ConstBinding:
      return n.initIsFunction ? `${name}()` : name;
    case NODE_KIND.LetBinding:
      return n.initIsFunction ? `${name}()` : `let ${name}`;
    case NODE_KIND.FormalParameter:
    case NODE_KIND.ReturnArgumentReference:
    case NODE_KIND.SyntheticIfStatementTest:
    case NODE_KIND.SyntheticSwitchStatementDiscriminant:
    case NODE_KIND.SyntheticWhileStatementTest:
    case NODE_KIND.SyntheticDoWhileStatementTest:
    case NODE_KIND.SyntheticForStatementHeader:
    case NODE_KIND.SyntheticForInStatementHeader:
    case NODE_KIND.SyntheticForOfStatementHeader:
    case NODE_KIND.SyntheticModuleSink:
    case NODE_KIND.SyntheticExpressionStatement:
      return name;
    case NODE_KIND.SyntheticBeyondDepth:
      // ASCII "..." rather than U+2026; some Mermaid renderers stumble on
      // multibyte glyphs inside a node-shape label.
      return "...";
  }
}

File: crates/unsnarl-emitter-mermaid/src/node_head.rs
//! Builds the leading text fragment of a node's mermaid label
//! (before the trailing `<br/>L<line>` range stamp).

use unsnarl_oxc_parity::VariableDeclarationKind;
use unsnarl_visual_graph::node_kind::NodeKind;
use unsnarl_visual_graph::visual_node::VisualNode;

use crate::escape::escape_into;

pub fn node_head(n: &VisualNode) -> String {
    let mut out = String::new();
    node_head_into(&mut out, n);
    out
}

/// Destination-arg variant of [`node_head`]: writes the head fragment
/// straight into `out`. Used by `emit_node` so the per-node label
/// build does not bounce through an intermediate `String` for the
/// "name", a second `String` for the head shell, and a third one for
/// the outer line.
pub fn node_head_into(out: &mut String, n: &VisualNode) {
    if n.is_jsx_element() {
        // Mermaid `["..."]` labels require HTML-escaped angle
        // brackets so the parser does not mistake them for syntax;
        // the renderer surfaces them as literal `<` / `>` in the
        // output.
        out.push_str("&lt;");
        escape_into(out, n.name());
        out.push_str("&gt;");
        return;
    }
    match n.kind() {
        NodeKind::FunctionDeclaration => {
            escape_into(out, n.name());
            out.push_str("()");
        }
        NodeKind::ClassDeclaration => {
            out.push_str("class ");
            escape_into(out, n.name());
        }
        NodeKind::NamedImportBinding => match n.imported_name() {
            Some(imported) if imported != n.name() => {
                escape_into(out, n.name());
            }
            _ => {
                out.push_str("import ");
                escape_into(out, n.name());
            }
        },
        NodeKind::DefaultImportBinding | NodeKind::NamespaceImportBinding => {
            out.push_str("import ");
            escape_into(out, n.name());
        }
        NodeKind::CatchParameter => {
            out.push_str("catch ");
            escape_into(out, n.name());
        }
        NodeKind::SyntheticImplicitGlobal => {
            out.push_str("global ");
            escape_into(out, n.name());
        }
        NodeKind::WriteReference => match n.declaration_kind() {
            Some(VariableDeclarationKind::Let) => {
                out.push_str("let ");
                escape_into(out, n.name());
            }
            _ => {
                escape_into(out, n.name());
            }
        },
        NodeKind::SyntheticModuleSource => {
            out.push_str("module ");
            escape_into(out, n.name());
        }
        NodeKind::SyntheticImportIntermediate => {
            out.push_str("import ");
            escape_into(out, n.name());
        }
        NodeKind::VarBinding => {
            if n.init_is_function() {
                escape_into(out, n.name());
                out.push_str("()");
            } else {
                out.push_str("var ");
                escape_into(out, n.name());
            }
        }
        NodeKind::ConstBinding => {
            if n.init_is_function() {
                escape_into(out, n.name());
                out.push_str("()");
            } else {
                escape_into(out, n.name());
            }
        }
        NodeKind::LetBinding => {
            if n.init_is_function() {
                escape_into(out, n.name());
                out.push_str("()");
            } else {
                out.push_str("let ");
                escape_into(out, n.name());
            }
        }
        NodeKind::UsingBinding => {
            if n.init_is_function() {
                escape_into(out, n.name());
                out.push_str("()");
            } else {
                out.push_str("using ");
                escape_into(out, n.name());
            }
        }
        NodeKind::AwaitUsingBinding => {
            if n.init_is_function() {
                escape_into(out, n.name());
                out.push_str("()");
            } else {
                out.push_str("await using ");
                escape_into(out, n.name());
            }
        }
        NodeKind::FormalParameter
        | NodeKind::ReturnArgumentReference
        | NodeKind::ThrowArgumentReference
        | NodeKind::SyntheticIfStatementTest
        | NodeKind::SyntheticSwitchStatementDiscriminant
        | NodeKind::SyntheticWhileStatementTest
        | NodeKind::SyntheticDoWhileStatementTest
        | NodeKind::SyntheticForStatementHeader
        | NodeKind::SyntheticForInStatementHeader
        | NodeKind::SyntheticForOfStatementHeader
        | NodeKind::SyntheticModuleSink
        | NodeKind::SyntheticExpressionStatement
        | NodeKind::SyntheticBreakStatement
        | NodeKind::SyntheticContinueStatement => {
            escape_into(out, n.name());
        }
        NodeKind::SyntheticBeyondDepth => {
            out.push_str("...");
        }
    }
}

#[cfg(test)]
#[path = "node_head_test.rs"]
mod node_head_test;
